java/arcs/core/util/LruCacheMap.kt

```diff
@@ -26,18 +29,21 @@ open class LruCacheMap<K, V>(
   val capacity: Int = 100,
   private val backingMap: LinkedHashMap<K, V> = linkedMapOf(),
   val livenessPredicate: ((K, V) -> Boolean) = { _, _ -> true },
+  val evictor: ((() -> Unit) -> Unit) = { it() },
   val onEvict: ((K, V) -> Unit)? = null
 ) : MutableMap<K, V> by backingMap {
 
+  private val pendingEviction = atomic(mapOf<K, V>())
```

Comment:
Can this just be a list of map entries, instead of a map?

Reply:
Do we need this internal container of pending evictions? 

Wouldn't it work to provide an implementation for `onEvict` that managed the implementations however it wanted to? 

i.e.

```
class MyEvictionQueue<K, V>(
  scope: CoroutineScope
) {
  val evictionChannel= Channel<Map.Entry<K,V>(UNLIMITED)

  val evictionJob = scope.launch {
    for entry in channel {
      // close, etc.
    }
  }

  fun enqueueEviction(entry: Map.Entry<K,V>) {
    channel.sendBlocking(entry)
  }
}
```

Then we just pass `myEvictor::enqueueEviction` as the `onEvict` method?

Reply:
I guess the central design question is, what guarantees to do want. 

In other words, what do we guarantee before onEvict() is called, what do we guarantee during its processing, and what do we guarantee afterwards.

The main issue is, if you immediately remove the backingMap item and run the onEvict, it's potentially closing the database/store. So the constrain really is, onEvict() is waiting on all writes to be flushed before closing the store, what happens in the meantime if someone tries to .get() the store and do operations, where the onEvict is potentially racing it. 

I mean, we could dramatically simplify this by making the LruCacheMap suspendable, adding a mutex, and preventing any operations until onEvict() returns. But that might also slow things down and become a bottleneck. 

I though of the use of a Channel like you did, but I need to be convinced we can prevent the race of an eviction method racing with someone creating a new store.

Reply:
I think the only guarantee we should provide is that at when `onEvict` is called, the item no longer exists in the LRUCache's map.

` if someone tries to .get() the store and do operations, where the onEvict is potentially racing it.` 

If we provide the above guarantee, the item has already have been removed from the LRUCache map, and will not be returned ever again for `.get` calls (which I think is already true for the existing implementation, right?)

`I need to be convinced we can prevent the race of an eviction method racing with someone creating a new store.`

If a store is evicted from the map, then a new store that's added to the map should be a new instance that we can use (and eventually close) independently of the one that was just removed.

With this, there's still the potential issue of a `put` racing a `get` resulting in the `get` returning something that was just evicted. But in our current usage, this won't happen I thing, because we always work with the map within a mutex.